File: Algorithm_108_Convert Sorted Array to Binary Search Tree.cpp
#include <iostream>
#include <vector>
using namespace std;
struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};
TreeNode* sortedArrayToBST(vector<int>& nums)
{
	if (nums.size()==0) return NULL;
	if (nums.size()==1) return new TreeNode(nums[0]);
	int mid=nums.size()/2;
	TreeNode *root=new TreeNode(nums[mid]);
	vector<int> leftNodes(nums.begin(),nums.begin()+mid);
	vector<int> rightNodes(nums.begin()+mid+1,nums.end());
	root->left=sortedArrayToBST(leftNodes);
	root->right=sortedArrayToBST(rightNodes);
	return root;
}

File: Algorithm_108_Convert Sorted Array to Binary Search Tree.java
class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode(int x) { val = x; }
}

// 思路：找有序数组的正中间的元素作为根节点，对左半部分和右半部分递归，分别作为左子树和右子树

class Solution {

    public TreeNode sortedArrayToBST(int[] nums) {
        if (nums == null || nums.length == 0) {
            return null;
        }
        return sortedArrayToBST(nums, 0, nums.length - 1);
    }

    private TreeNode sortedArrayToBST(int[] nums, int left, int right) {
        if (left > right) {
            return null;
        } else if (left == right) {
            return new TreeNode(nums[left]);
        }
        int mid = (left + right) / 2;
        TreeNode root = new TreeNode(nums[mid]);
        root.left = sortedArrayToBST(nums, left, mid - 1);
        root.right = sortedArrayToBST(nums, mid + 1, right);
        return root;
    }

}